metabox/metabox/core/machine.py

```diff
@@ -240,57 +240,88 @@ def switch_on_networking(self):
         return run_or_raise(self._container, "sudo ip link set eth0 up")
 
 
-class ContainerVenvMachine(ContainerBaseMachine):
+class ContainerSourceMachine(ContainerBaseMachine):
     """
-    Machine using LXD container as the backend and running checkbox
-    in a virtualenv.
+    Machine using LXD container as the backend and running checkbox from
+    source code repository.
     """
 
     def __init__(self, config, container):
         super().__init__(config, container)
-        if self.config.role == 'service':
-            self._checkbox_wrapper = 'sudo /home/ubuntu/run.sh {}'.format(
-                self.CHECKBOX)
-        else:
-            self._checkbox_wrapper = '/home/ubuntu/run.sh {}'.format(
-                self.CHECKBOX)
         self._pts = None  # Keep a pointer to started pts for easy kill
 
     def get_early_dir_transfer(self):
-        return [(self.config.uri, '/home/ubuntu/checkbox-ng')]
+        dirs = [
+            (self.config.uri, "/home/ubuntu/checkbox"),
+            (Path(self.config.uri) / "providers/base",
+             "/var/tmp/checkbox-providers/base"),
+            (Path(self.config.uri) / "providers/resource",
+             "/var/tmp/checkbox-providers/resource"),
+            (Path(self.config.uri) / "providers/certification-client",
+             "/var/tmp/checkbox-providers/certification-client"),
+        ]
+        return dirs
 
     def get_early_setup(self):
-        """Virtualenv creation."""
-        return [
-            '''bash -c "printf '#!/bin/bash\\n. '''
-            '''/home/ubuntu/checkbox-ng/venv/bin/activate\\n$@' > '''
-            '''/home/ubuntu/run.sh"''',
-            'chmod +x /home/ubuntu/run.sh',
-            'chmod +x /home/ubuntu/checkbox-ng/setup.py',
-            'chmod +x /home/ubuntu/checkbox-ng/mk-venv',
-            "bash -c 'pushd /home/ubuntu/checkbox-ng ; ./mk-venv venv'",
+        """
+        Installation from source and, if required, creation of systemd service.
+        """
+        service_file = ("[Unit]\\n"
```

Comment:
How do you feel about having a multiline string here that's dedented afterwards?

Reply:
Solved in commit [2a27407](https://github.com/canonical/checkbox/pull/305/commits/2a27407fff2090a5c22624c67551bed6488df917)